### PR TITLE
smoke b2c user control functionality

### PR DESCRIPTION
**Pull request by khushal-rathore-metacube**
https://spryker.atlassian.net/browse/CC-21838

Performed smoke b2c user control scenario:

1. Create a new Role / a new Group, assign to a user, check in Zed for the user
2. Create a new user + login in Zed
3. Delete / Deactivate a user
4. Edit a user

### CODE REVIEW

File: tests/smoke/smoke_b2c.robot [hunk 1]
```diff
@@ -599,3 +600,31 @@ Refunds
...
+    Zed: add a new user detatils:    abc${random}@gmail.com   change${random}   change${random}    khushal    singh
+    Zed: logout
+    Zed: login on Zed with provided credentials:    abc${random}@gmail.com    change${random}
+    Zed: go to second navigation item level:    Users    User Roles
+    Zed: click button in Header:    Add new Role
+    Zed: Add role name:    role${random}
+    Zed: submit the form
+    Zed: go to second navigation item level:    Users    User Groups
+    Zed: click button in Header:    Create Group
+    Zed: add new group:   Group${random}
```
Comment (AntonShopin) [inline tests/smoke/smoke_b2c.robot:616]: see comment above from keywrods

File: tests/smoke/smoke_b2c.robot [hunk 2]
```diff
@@ -599,3 +600,31 @@ Refunds
...
+    Zed: perform search by:    abc${random}@gmail.com
+    Zed: edit user Details
+    Zed: Check checkbox by Label:    Group${random}
+    Zed: submit the form
+    Zed: message should be shown:    User was updated successfully.
+    Zed: logout
+    Zed: login on Zed with provided credentials:    ${zed_admin_email}
+    Zed: go to second navigation item level:    Users    Users
+    Zed: perform search by:    abc${random}@gmail.com
+    Zed: deactivate user
```
Comment (AntonShopin) [inline tests/smoke/smoke_b2c.robot:627]: if you deactivate the user, how do you test that it's deactivated? Maybe it's better to try to log in with it and see that you can't ?

**Review by AntonShopin — CHANGES_REQUESTED** [review]
You need to cover the main idea of the user control feature - different roles. 
Please check this article: https://docs.spryker.com/docs/scos/user/back-office-user-guides/202204.0/users/best-practices-managing-users-and-their-permissions-with-roles-and-groups.html#creating-the-role-with-limited-access

With your tests, you need to create a new role with limited access and add a new group and user. Assign the user to it and log in with this user -> check that this user is not able to access something in Zed (pages or actions). 

To do this - you need to create a new keyword to manage different permissions in the role

File: tests/smoke/smoke_b2c.robot [hunk 3]
```diff
@@ -599,3 +600,31 @@ Refunds
     [Teardown]    Run keywords    Zed: login on Zed with provided credentials:    ${zed_admin_email}
     ...    AND    Zed: activate following discounts from Overview page:    Tu & Wed $5 off 5 or more    10% off $100+    20% off cameras    Tu & Wed €5 off 5 or more    10% off minimum order
  
+User_control_functionality
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply tests/smoke/smoke_b2c.robot:603]: Done

File: resources/steps/zed_users_steps.robot [hunk 4]
```diff
@@ -13,4 +14,29 @@ Zed: delete Zed user with the following email:
...
+Zed: add a new user detatils:
+    [Arguments]     ${user_name}    ${password}    ${confirm_password}    ${name}    ${last_name}
+    Type Text    ${zed_user_email_field}   ${user_name}
+    Type Text    ${zed_user_password_filed}    ${password}
+    Type Text    ${zed_user_repeat_password_field}   ${confirm_password}
+    Type Text    ${zed_user_first_name_field}    ${name}
+    Type Text    ${zed_user_last_name_field}    ${last_name}
+    Zed: Check checkbox by Label:    Root group
+    Select From List By Label    ${zed_user_interface_language}    en_US
+    Zed: submit the form
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply resources/steps/zed_users_steps.robot:26]: Done

File: resources/pages/zed/zed_create_zed_user_page.robot [hunk 5]
```diff
@@ -5,3 +5,8 @@ ${zed_user_repeat_password_field}    id=user_password_second
 ${zed_user_first_name_field}    id=user_first_name
 ${zed_user_last_name_field}    id=user_last_name
 ${zed_user_interface_language}    id=user_fk_locale
+${zed_user_role_name}    id=role_name
+${zed_user_group_title}     id=group_title
+${zed_user_group_assigned_role_textbox}    xpath=//input[@role='searchbox']
+${zed_user_edit_button}    xpath=(//a[contains(@class,'btn-edit')])[2]
+${zed_user_deactivate_button}    xpath=//form[contains(@name,'deactivate_user_form')]
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply resources/pages/zed/zed_create_zed_user_page.robot]: Done

File: resources/steps/zed_users_steps.robot [hunk 6]
```diff
@@ -13,4 +14,29 @@ Zed: delete Zed user with the following email:
...
 
+Zed: Add role name:
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply resources/steps/zed_users_steps.robot]: Done

File: resources/steps/zed_users_steps.robot [hunk 7]
```diff
@@ -13,4 +14,29 @@ Zed: delete Zed user with the following email:
...
+    [Arguments]    ${name}
+     Type Text    ${zed_user_role_name}      ${name}
+
+Zed: add new group:
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply resources/steps/zed_users_steps.robot]: Done

File: resources/steps/zed_users_steps.robot [hunk 8]
```diff
@@ -13,4 +14,29 @@ Zed: delete Zed user with the following email:
...
+     [Arguments]    ${Group_name}
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply resources/steps/zed_users_steps.robot]: Done

File: tests/smoke/smoke_b2c.robot [hunk 9]
```diff
@@ -599,3 +600,31 @@ Refunds
...
+    Zed: login on Zed with provided credentials:    ${zed_admin_email}
+    Zed: go to second navigation item level:    Users    Users
+    Zed: click button in Header:    Add New User
+    Zed: add a new user detatils:    abc${random}@gmail.com   change${random}   change${random}    khushal    singh
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply tests/smoke/smoke_b2c.robot]: Done

File: tests/smoke/smoke_b2c.robot (hunk 1 above)
Comment (khushal-rathore-metacube) [reply tests/smoke/smoke_b2c.robot]: Done

File: tests/smoke/smoke_b2c.robot (hunk 2 above)
Comment (khushal-rathore-metacube) [reply tests/smoke/smoke_b2c.robot]: Done

File: resources/steps/zed_users_steps.robot [hunk 10]
```diff
@@ -13,4 +17,57 @@ Zed: delete Zed user with the following email:
...
+    [Arguments]    ${name}    ${full_access}=*
+    Zed: go to second navigation item level:    Users    User Roles
+    Zed: click button in Header:    Add new Role
+    Type Text    ${zed_user_role_name}      ${name}
+    Zed: submit the form
+    Zed: message should be shown:    success    Role "${name}" successfully added.
+    Type Text    ${zed_user_role_bundle_locator}     ${full_access}
+    Type Text    ${zed_user_role_controller_locator}    ${full_access}
+    Type Text    ${zed_user_role_action_locator}    ${full_access}
+    Click    ${zed_add_rule_button}
```
Comment (AntonShopin) [inline resources/steps/zed_users_steps.robot:30]: Why these steps are here, if you have another keyword for it? `Zed: apply restrictions for user role:`

File: resources/steps/zed_users_steps.robot [hunk 11]
```diff
@@ -13,4 +17,57 @@ Zed: delete Zed user with the following email:
...
+    
+Zed: apply restrictions for user role:
+    [Arguments]    ${bundle_access}    ${controller_access}    ${action_access}    ${permission}
```
Comment (AntonShopin) [inline resources/steps/zed_users_steps.robot:33]: There are no conflicts? You have added new variables in the environment file - cool, I appreciate it. But in this keyword, you use exactly the same naming for arguments (which are also some kind of variables). Maybe it's better to rename argument so the are not mixed with variables from the env file?

File: resources/common/common_zed.robot [hunk 12]
```diff
@@ -114,8 +114,13 @@ Zed: go to tab:
...
 Zed: message should be shown:
-    [Arguments]    ${text}
-    Wait Until Element Is Visible    xpath=//div[contains(@class,'alert alert-success')]//*[contains(text(),'${text}')]    message=Success message is not shown
+    [Arguments]    ${type}    ${text}
+    IF    '${type}' == 'success'
+        Wait Until Element Is Visible    xpath=//div[contains(@class,'alert alert-success')]//*[contains(text(),'${text}')]    message=Success message is not shown
+    ELSE IF    '${type}' == 'fail'
+        Wait Until Element Is Visible    xpath=//div[contains(@class,'alert alert-danger')]//*[contains(text(),'${text}')]    message=Success message is not shown
+    END
+
```
Comment (AntonShopin) [inline resources/common/common_zed.robot:123]: are you sure that you have updated ALL tests and ALL keywords for ALL demoshops, where this keyword was used before? 
Previously it was required to provide only one argument. Now, with your changes, this keyword will run only if 2 arguments have been provided. 

For me it's a BC break, not allowed OR - you need to update ALL other places where this keyword is already used.

File: resources/steps/zed_users_steps.robot [hunk 13]
```diff
@@ -13,4 +17,57 @@ Zed: delete Zed user with the following email:
     Wait Until Page Contains Element    ${zed_confirm_delete_user_button}
     Click    ${zed_confirm_delete_user_button}
 
+Zed: create new role with name:
+    [Arguments]    ${name}    ${full_access}=*
+    Zed: go to second navigation item level:    Users    User Roles
+    Zed: click button in Header:    Add new Role
+    Type Text    ${zed_user_role_name}      ${name}
+    Zed: submit the form
+    Zed: message should be shown:    success    Role "${name}" successfully added.
```
Comment (AntonShopin) [inline resources/steps/zed_users_steps.robot:26]: see my comment above about your changes to this keyword

File: resources/steps/zed_users_steps.robot [hunk 14]
```diff
@@ -13,4 +17,57 @@ Zed: delete Zed user with the following email:
...
+    [Arguments]    ${bundle_access}    ${controller_access}    ${action_access}    ${permission}
+    Type Text    ${zed_user_role_bundle_locator}     ${bundle_access}
+    Type Text    ${zed_user_role_controller_locator}    ${controller_access}
+    Type Text    ${zed_user_role_action_locator}    ${action_access}
+    Select From List By Label    ${zed_user_role_permission}    ${permission}
+    Click    ${zed_add_rule_button}
+    
 
+Zed: create new group with role assigned:
+    [Arguments]    ${Group_name}    ${role_name}
```
Comment (AntonShopin) [inline resources/steps/zed_users_steps.robot:42]: Group_name should be in lower case

File: resources/steps/zed_users_steps.robot [hunk 15]
```diff
@@ -13,4 +17,57 @@ Zed: delete Zed user with the following email:
...
+    Element Text Should Be    ${attribute_access_denied_header}    ${message}   
+
+Zed: deactivate the created user:
+    [Arguments]    ${email}
+    Zed: go to second navigation item level:    Users    Users
+    Zed: click Action Button in a table for row that contains:    ${email}    Deactivate
+    Zed: message should be shown:    success   User was deactivated successfully.
+
+Zed: add the created group to created user:
+    [Arguments]    ${Group_name}
```
Comment (AntonShopin) [inline resources/steps/zed_users_steps.robot:62]: Group_name should be in lower case

File: resources/steps/zed_users_steps.robot [hunk 16]
```diff
@@ -13,4 +17,57 @@ Zed: delete Zed user with the following email:
...
+    Zed: go to second navigation item level:    Users    Users
+    Zed: click Action Button in a table for row that contains:    abc${random}@gmail.com    Edit
```
Comment (AntonShopin) [inline resources/steps/zed_users_steps.robot:64]: from where did you get `abc${random}@gmail.com` ? Hardcode, no? Please refactor

File: tests/smoke/smoke_b2c.robot [hunk 17]
```diff
@@ -1,5 +1,6 @@
 *** Settings ***
 Library    BuiltIn
+Library    XML
```
Comment (AntonShopin) [inline tests/smoke/smoke_b2c.robot:3]: what is this? From where and why do you need XML library?

File: tests/smoke/smoke_b2c.robot [hunk 18]
```diff
@@ -599,3 +601,19 @@ Refunds
     [Teardown]    Run keywords    Zed: login on Zed with provided credentials:    ${zed_admin_email}
     ...    AND    Zed: activate following discounts from Overview page:    Tu & Wed $5 off 5 or more    10% off $100+    20% off cameras    Tu & Wed €5 off 5 or more    10% off minimum order
  
+User_Control
+    [Documentation]    Create a user with role having limited access and group assigned to it,deactivate the user as well.
+    Zed: login on Zed with provided credentials:    ${zed_admin_email}
+    Zed: create new Zed user with the following data:    abc${random}@gmail.com   change${random}    Test    last    Root group    This user is an agent    en_US
```
Comment (AntonShopin) [inline tests/smoke/smoke_b2c.robot:607]: Why first not to:
- Create new role
- Apply restrictions to this role
- Create new Group
- And only then create new user with this new group?

File: resources/pages/zed/zed_user_role_page.robot [hunk 19]
```diff
@@ -0,0 +1,7 @@
+*** Variables ***
+${zed_user_role_name}    id=role_name
+${zed_user_role_bundle_locator}    id=ruleset_bundle
+${zed_user_role_controller_locator}    id=ruleset_controller
+${zed_user_role_action_locator}    id=ruleset_action
+${zed_user_role_permission}    id=ruleset_type
+${zed_add_rule_button}    xpath=//form[@name='ruleset']//input[@class='btn btn-primary safe-submit']
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply resources/pages/zed/zed_user_role_page.robot:7]: Fixed

File: resources/steps/zed_users_steps.robot (hunk 10 above)
Comment (khushal-rathore-metacube) [reply resources/steps/zed_users_steps.robot:30]: fixed

File: resources/steps/zed_users_steps.robot (hunk 11 above)
Comment (khushal-rathore-metacube) [reply resources/steps/zed_users_steps.robot:33]: fixed

File: resources/common/common_zed.robot (hunk 12 above)
Comment (khushal-rathore-metacube) [reply resources/common/common_zed.robot:123]: fixed

File: resources/steps/zed_users_steps.robot (hunk 13 above)
Comment (khushal-rathore-metacube) [reply resources/steps/zed_users_steps.robot:26]: fixed

File: resources/steps/zed_users_steps.robot (hunk 14 above)
Comment (khushal-rathore-metacube) [reply resources/steps/zed_users_steps.robot:42]: fixed

File: resources/steps/zed_users_steps.robot (hunk 15 above)
Comment (khushal-rathore-metacube) [reply resources/steps/zed_users_steps.robot:62]: fixed

File: resources/pages/zed/zed_catalog_attributes_page.robot [hunk 20]
```diff
@@ -0,0 +1,2 @@
+*** Variables ***
+${attribute_access_denied_header}    xpath=//div[@class='wrapper wrapper-content']//h1
```
[the remark this answers is not in the document]
Comment (khushal-rathore-metacube) [reply resources/pages/zed/zed_catalog_attributes_page.robot:2]: Fixed

File: tests/smoke/smoke_b2c.robot (hunk 18 above)
Comment (khushal-rathore-metacube) [reply tests/smoke/smoke_b2c.robot:607]: Fixed

File: tests/smoke/smoke_b2c.robot (hunk 17 above)
Comment (khushal-rathore-metacube) [reply tests/smoke/smoke_b2c.robot:3]: Fixed

File: resources/steps/zed_users_steps.robot (hunk 16 above)
Comment (khushal-rathore-metacube) [reply resources/steps/zed_users_steps.robot:64]: Fixed

File: resources/steps/zed_users_steps.robot (hunk 10 above)
Comment (AntonShopin) [reply resources/steps/zed_users_steps.robot]: not fixed. My comment was to remove it from this keyword and you have just replaced inline action with a keyword. 

You need to design your test using different available keywords. Don't put everything under one keyword. 
Like, with keyword one you 'create a new role', using second keyword in your test - you specify restrictions

File: tests/smoke/smoke_b2c.robot [hunk 21]
```diff
@@ -599,3 +600,19 @@ Refunds
     [Teardown]    Run keywords    Zed: login on Zed with provided credentials:    ${zed_admin_email}
     ...    AND    Zed: activate following discounts from Overview page:    Tu & Wed $5 off 5 or more    10% off $100+    20% off cameras    Tu & Wed €5 off 5 or more    10% off minimum order
  
+User_Control
+    [Documentation]    Create a user with role having limited access and group assigned to it,deactivate the user as well.
+    Zed: login on Zed with provided credentials:    ${zed_admin_email}
+    Zed: create new role with name:    role${random}
+    Zed: apply restrictions for user role:    ${bundle_access}    ${controller_access}    ${action_access}    ${permission_deny}
```
Comment (AntonShopin) [inline tests/smoke/smoke_b2c.robot:607]: Check my comment for `Zed: create new role with name:` keyword, where you already, for some reason, use `Zed: apply restrictions for user role:`
You see, we can create new roles with everything denied, with you hardcoded keyword in keyword all new roles will be created with 'all allowed', not good

Comment (khushal-rathore-metacube) [reply tests/smoke/smoke_b2c.robot:607]: Fixed. created two keywords `Zed: create new role with name:`  and `Zed: apply access permissions for user role:`

File: resources/steps/zed_users_steps.robot (hunk 10 above)
Comment (khushal-rathore-metacube) [reply resources/steps/zed_users_steps.robot]: Fixed

**Pull request closed without merging**